syntax = "proto3";

package distributed_printing;

import "google/protobuf/empty.proto";

// Serviço para o servidor de impressão BURRO (implementado no servidor)
service PrintingService {
  rpc SendToPrinter (PrintRequest) returns (PrintResponse);
}

// Serviço para comunicação entre CLIENTES (implementado nos clientes)
service MutualExclusionService {
  rpc RequestAccess (AccessRequest) returns (AccessResponse);
  rpc ReleaseAccess (AccessRelease) returns (google.protobuf.Empty);
}

// Mensagens para impressão (cliente -> servidor burro)
message PrintRequest {
  int32 client_id = 1;
  string message_content = 2;
  int64 lamport_timestamp = 3;
  int32 request_number = 4;
}

message PrintResponse {
  bool success = 1;
  string confirmation_message = 2;
  int64 lamport_timestamp = 3;
}

// Mensagens para exclusão mútua (cliente <-> cliente)
message AccessRequest {
  int32 client_id = 1;
  int64 lamport_timestamp = 2;
  int32 request_number = 3;
}

message AccessResponse {
  bool access_granted = 1;
  int64 lamport_timestamp = 2;
}

message AccessRelease {
  int32 client_id = 1;
  int64 lamport_timestamp = 2;
  int32 request_number = 3;
}
